File: clutter/lib/entities/photo.dart
import 'timestamps.dart';
import 'topo.dart';
import 'types.dart';
import 'upload.dart';

class Photo {
  final int id;
  final String title;
  final String? description;
  final Upload? upload;
  final Timestamps timestamps;
  final Topo? topo;

  const Photo({
    required this.id,
    required this.title,
    this.description,
    this.upload,
    required this.timestamps,
    this.topo,
  });

  factory Photo.fromJson(JsonObject json) {
    if (json
        case {
          'id': int id,
          'title': String title,
          'description': String? description,
          'upload': JsonObject upload,
          'topo': JsonObject? topo,
        }) {
      return Photo(
        id: id,
        title: title,
        description: description,
        upload: Upload.fromJson(upload),
        timestamps: Timestamps.fromJson(json),
        topo: topo != null ? Topo.fromJson(topo) : null,
      );
    }
    throw JSONException("Photo", json);
  }

  JsonObject toJson() {
    return {
      'id': id,
      'title': title,
      'description': description,
      'upload': upload?.toJson(),
      ...timestamps.toJson(),
    };
  }
}

File: go-server/internal/models/photo.go
package models

type Photo struct {
	ID          int64      `json:"id" sqlp:"id"`
	Title       string     `json:"title" sqlp:"title"`
	Description *string    `json:"description,omitzero" sqlp:"description"`
	Upload      *Upload    `json:"upload,omitzero" sqlp:"upload"`
	Photoable   *Photoable `json:"photoable,omitzero" sqlp:"photoable"`
	Topo        *Topo      `json:"topo,omitzero" sqlp:"topo"`
	Timestamps
}

func (p Photo) IsZero() bool {
	return p.ID == 0 &&
		p.Title == "" &&
		p.Description == nil &&
		p.Upload == nil &&
		p.Photoable == nil &&
		p.Topo == nil &&
		p.Timestamps.IsZero()
}
